File: android/src/main/java/com/getcapacitor/community/intercom/IntercomPlugin.java
package com.getcapacitor.community.intercom;

import androidx.annotation.NonNull;

import com.getcapacitor.CapConfig;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

import io.intercom.android.sdk.Intercom;
import io.intercom.android.sdk.IntercomContent;
import io.intercom.android.sdk.IntercomError;
import io.intercom.android.sdk.IntercomSpace;
import io.intercom.android.sdk.IntercomStatusCallback;
import io.intercom.android.sdk.UnreadConversationCountListener;
import io.intercom.android.sdk.UserAttributes;
import io.intercom.android.sdk.identity.Registration;
import io.intercom.android.sdk.push.IntercomPushClient;

@CapacitorPlugin(name = "Intercom", permissions = @Permission(strings = {}, alias = "receive"))
public class IntercomPlugin extends Plugin implements UnreadConversationCountListener {
    private final IntercomPushClient intercomPushClient = new IntercomPushClient();

    @Override
    public void load() {
        // Set up Intercom
        setUpIntercom();

        // load parent
        super.load();
    }

    @Override
    public void handleOnStart() {
        super.handleOnStart();
        bridge.getActivity().runOnUiThread(() -> {
            //We also initialize intercom here just in case it has died. If Intercom is already set up, this won't do anything.
            setUpIntercom();
            Intercom.client().handlePushMessage();
        });
    }

    @PluginMethod
    @Deprecated
    public void registerIdentifiedUser(PluginCall call) {
        String email = call.getString("email");
        String userId = call.getData().getString("userId");

        Registration registration = new Registration();

        if (email != null && email.length() > 0) {
            registration = registration.withEmail(email);
        }
        if (userId != null && userId.length() > 0) {
            registration = registration.withUserId(userId);
        }
        Intercom.client().loginIdentifiedUser(registration, new IntercomStatusCallback() {
            @Override
            public void onSuccess() {
                call.resolve();
            }

            @Override
            public void onFailure(@NonNull IntercomError intercomError) {
                call.reject("Error logging in: " + intercomError.getErrorMessage());
            }
        });
    }

    @PluginMethod
    public void loginIdentifiedUser(PluginCall call) {
        String email = call.getString("email");
        String userId = call.getData().getString("userId");

        Registration registration = new Registration();

        if (email != null && email.length() > 0) {
            registration = registration.withEmail(email);
        }
        if (userId != null && userId.length() > 0) {
            registration = registration.withUserId(userId);
        }
        Intercom.client().loginIdentifiedUser(registration, new IntercomStatusCallback() {
            @Override
            public void onSuccess() {
                call.resolve();
            }

            @Override
            public void onFailure(@NonNull IntercomError intercomError) {
                call.reject("Error logging in: " + intercomError.getErrorMessage());
            }
        });
    }

    @PluginMethod
    @Deprecated
    public void registerUnidentifiedUser(PluginCall call) {
        Intercom.client().loginUnidentifiedUser(new IntercomStatusCallback() {
            @Override
            public void onSuccess() {
                call.resolve();
            }

            @Override
            public void onFailure(@NonNull IntercomError intercomError) {
                call.reject("Error logging in unidentified user: " + intercomError.getErrorMessage());
            }
        });
    }

    @PluginMethod
    public void loginUnidentifiedUser(PluginCall call) {
        Intercom.client().loginUnidentifiedUser(new IntercomStatusCallback() {
            @Override
            public void onSuccess() {
                call.resolve();
            }

            @Override
            public void onFailure(@NonNull IntercomError intercomError) {
                call.reject("Error logging in unidentified user: " + intercomError.getErrorMessage());
            }
        });
    }

    @PluginMethod
    public void updateUser(PluginCall call) {
        UserAttributes.Builder builder = new UserAttributes.Builder();
        String userId = call.getString("userId");
        if (userId != null && userId.length() > 0) {
            builder.withUserId(userId);
        }
        String email = call.getString("email");
        if (email != null && email.length() > 0) {
            builder.withEmail(email);
        }
        String name = call.getString("name");
        if (name != null && name.length() > 0) {
            builder.withName(name);
        }
        String phone = call.getString("phone");
        if (phone != null && phone.length() > 0) {
            builder.withPhone(phone);
        }
        String languageOverride = call.getString("languageOverride");
        if (languageOverride != null && languageOverride.length() > 0) {
            builder.withLanguageOverride(languageOverride);
        }
        Map<String, Object> customAttributes = mapFromJSON(call.getObject("customAttributes"));
        if (customAttributes != null) {
            builder.withCustomAttributes(customAttributes);
        }
        Intercom.client().updateUser(builder.build(), new IntercomStatusCallback() {
            @Override
            public void onSuccess() {
                call.resolve();
            }

            @Override
            public void onFailure(@NonNull IntercomError intercomError) {
                call.reject("Error updating user: " + intercomError.getErrorMessage());
            }
        });
    }

    @PluginMethod
    public void logout(PluginCall call) {
        Intercom.client().logout();
        call.resolve();
    }

    @PluginMethod
    public void logEvent(PluginCall call) {
        String eventName = call.getString("name", "");
        if (eventName == null || eventName.isEmpty()) {
            call.reject("Event name is missing or empty");
            return;
        }

        Map<String, Object> metaData = mapFromJSON(call.getObject("data"));
        if (metaData == null) {
            Intercom.client().logEvent(eventName);
        } else {
            Intercom.client().logEvent(eventName, metaData);
        }

        call.resolve();
    }

    @PluginMethod
    public void present(PluginCall call) {
        Map<String, IntercomSpace> spaceMapping = new HashMap<>();
        spaceMapping.put("helpCenter", IntercomSpace.HelpCenter);
        spaceMapping.put("messages", IntercomSpace.Messages);
        spaceMapping.put("home", IntercomSpace.Home);

        String spaceString = call.getString("space", "");
        IntercomSpace space = spaceMapping.get(spaceString);
        if (space == null) {
            space = IntercomSpace.Home;
        }

        Intercom.client().present(space);
        call.resolve();
    }


    @PluginMethod
    @Deprecated
    public void displayMessenger(PluginCall call) {
        Intercom.client().present();
        call.resolve();
    }

    @PluginMethod
    public void displayMessageComposer(PluginCall call) {
        String message = call.getString("message");
        Intercom.client().displayMessageComposer(message);
        call.resolve();
    }

    @PluginMethod
    @Deprecated
    public void displayHelpCenter(PluginCall call) {
        Intercom.client().present(IntercomSpace.HelpCenter);
        call.resolve();
    }

    @PluginMethod
    public void hideMessenger(PluginCall call) {
        Intercom.client().hideIntercom();
        call.resolve();
    }

    @PluginMethod
    public void displayLauncher(PluginCall call) {
        Intercom.client().setLauncherVisibility(Intercom.VISIBLE);
        call.resolve();
    }

    @PluginMethod
    public void hideLauncher(PluginCall call) {
        Intercom.client().setLauncherVisibility(Intercom.GONE);
        call.resolve();
    }

    @PluginMethod
    public void displayInAppMessages(PluginCall call) {
        Intercom.client().setInAppMessageVisibility(Intercom.VISIBLE);
        call.resolve();
    }

    @PluginMethod
    public void hideInAppMessages(PluginCall call) {
        Intercom.client().setLauncherVisibility(Intercom.GONE);
        call.resolve();
    }

    @PluginMethod
    public void presentContent(PluginCall call) {
        String contentId = call.getString("contentId");
        if (contentId == null || contentId.isEmpty()) {
            call.reject("contentId is missing or empty");
            return;
        }

        Map<String, IntercomContent> contentTypeMapping = new HashMap<>();
        contentTypeMapping.put("carousel", new IntercomContent.Carousel(contentId));
        contentTypeMapping.put("survey", new IntercomContent.Survey(contentId));
        contentTypeMapping.put("article", new IntercomContent.Article(contentId));

        String contentTypeString = call.getString("contentType", "");
        IntercomContent contentType = contentTypeMapping.get(contentTypeString);
        if (contentType == null) {
            call.reject("contentType not found");
            return;
        }

        Intercom.client().presentContent(contentType);
        call.resolve();
    }

    @PluginMethod
    @Deprecated
    public void displayCarousel(PluginCall call) {
        String carouselId = call.getString("carouselId");
        if (carouselId == null || carouselId.isEmpty()) {
            call.reject("carouselId is missing or empty");
            return;
        }
        Intercom.client().presentContent(new IntercomContent.Carousel(carouselId));
        call.resolve();
    }

    @PluginMethod
    @Deprecated
    public void displayArticle(PluginCall call) {
        String articleId = call.getString("articleId");
        if (articleId == null || articleId.isEmpty()) {
            call.reject("articleId is missing or empty");
            return;
        }
        Intercom.client().presentContent(new IntercomContent.Article(articleId));
        call.resolve();
    }

    @PluginMethod
    public void setUserHash(PluginCall call) {
        String hmac = call.getString("hmac");
        if (hmac == null || hmac.isEmpty()) {
            call.reject("hmac is missing or empty");
            return;
        }
        Intercom.client().setUserHash(hmac);
        call.resolve();
    }

    @PluginMethod
    public void setBottomPadding(PluginCall call) {
        String stringValue = call.getString("value");
        if (stringValue == null || stringValue.isEmpty()) {
            call.reject("value is missing or empty");
            return;
        }
        int value = Integer.parseInt(stringValue);
        Intercom.client().setBottomPadding(value);
        call.resolve();
    }

    @PluginMethod
    public void getUnreadConversationCount(PluginCall call) {
        Integer unreadCount = Intercom.client().getUnreadConversationCount();
        JSObject ret = new JSObject();
        ret.put("unreadCount", unreadCount);
        call.resolve(ret);
    }

    @PluginMethod
    public void sendPushTokenToIntercom(PluginCall call) {
        String token = call.getString("value");
        if (token == null || token.isEmpty()) {
            call.reject("value is missing or empty");
            return;
        }
        try {
            intercomPushClient.sendTokenToIntercom(this.getActivity().getApplication(), token);
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to send push token to Intercom", e);
        }
    }

    @PluginMethod
    public void receivePush(PluginCall call) {
        try {
            JSObject notificationData = call.getData();
            Map message = mapFromJSON(notificationData);
            if (intercomPushClient.isIntercomPush(message)) {
                intercomPushClient.handlePush(this.getActivity().getApplication(), message);
                call.resolve();
            } else {
                call.reject("Notification data was not a valid Intercom push message");
            }
        } catch (Exception e) {
            call.reject("Failed to handle received Intercom push", e);
        }
    }

    @PluginMethod
    public void setupUnreadConversationListener(PluginCall call) {
        Intercom.client().addUnreadConversationCountListener(this);
        call.resolve();
    }

    @PluginMethod
    public void removeUnreadConversationListener(PluginCall call) {
        Intercom.client().removeUnreadConversationCountListener(this);
        call.resolve();
    }

    @Override
    public void onCountUpdate(int i) {
        JSObject ret = new JSObject();
        ret.put("unreadCount", i);
        notifyListeners("updateUnreadCount", ret);
    }

    private void setUpIntercom() {
        try {
            // get config
            CapConfig config = this.bridge.getConfig();
            String apiKey = config.getPluginConfiguration("Intercom").getString("androidApiKey");
            String appId = config.getPluginConfiguration("Intercom").getString("androidAppId");

            // init intercom sdk
            Intercom.initialize(this.getActivity().getApplication(), apiKey, appId);
        } catch (Exception e) {
            Logger.error("Intercom", "ERROR: Something went wrong when initializing Intercom. Check your configurations", e);
        }
    }

    private static Map<String, Object> mapFromJSON(JSObject jsonObject) {
        if (jsonObject == null) {
            return null;
        }
        Map<String, Object> map = new HashMap<>();
        Iterator<String> keysIter = jsonObject.keys();
        while (keysIter.hasNext()) {
            String key = keysIter.next();
            Object value = getObject(jsonObject.opt(key));
            if (value != null) {
                map.put(key, value);
            }
        }
        return map;
    }

    private static Object getObject(Object value) {
        if (value instanceof JSObject) {
            value = mapFromJSON((JSObject) value);
        } else if (value instanceof JSArray) {
            value = listFromJSON((JSArray) value);
        }
        return value;
    }

    private static List<Object> listFromJSON(JSArray jsonArray) {
        List<Object> list = new ArrayList<>();
        for (int i = 0, count = jsonArray.length(); i < count; i++) {
            Object value = getObject(jsonArray.opt(i));
            if (value != null) {
                list.add(value);
            }
        }
        return list;
    }
}

File: ios/Plugin/IntercomPlugin.swift
import Foundation
import Capacitor
import Intercom

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(IntercomPlugin)
public class IntercomPlugin: CAPPlugin {
    public override func load() {
        let apiKey = getConfig().getString("iosApiKey") ?? "ADD_IN_CAPACITOR_CONFIG_JSON"
        let appId = getConfig().getString("iosAppId") ?? "ADD_IN_CAPACITOR_CONFIG_JSON"
        Intercom.setApiKey(apiKey, forAppId: appId)
        
#if DEBUG
        Intercom.enableLogging()
#endif
        
        NotificationCenter.default.addObserver(self, selector: #selector(self.didRegisterWithToken(notification:)), name: Notification.Name.capacitorDidRegisterForRemoteNotifications, object: nil)
    }
    
    
    @objc func didRegisterWithToken(notification: NSNotification) {
        guard let deviceToken = notification.object as? Data else {
            return
        }
        Intercom.setDeviceToken(deviceToken)
    }
    
    @objc func setupUnreadConversationListener(_ call: CAPPluginCall) {
        NotificationCenter.default.addObserver(self,
           selector: #selector(self.updateUnreadCount(notification:)),
               name: NSNotification.Name.IntercomUnreadConversationCountDidChange,
             object: nil)
    }
    
    @objc func updateUnreadCount(notification: NSNotification) {
        guard let unreadCount = notification.object as? Data else {
            return
        }
        
        notifyListeners("updateUnreadCount", data: ["unreadCount": unreadCount])
    }
    
    @available(*, deprecated, message: "This method is deprecated, use loginIdentifiedUser() instead.")
    @objc func registerIdentifiedUser(_ call: CAPPluginCall) {
        let userId = call.getString("userId")
        let email = call.getString("email")
        let attributes = ICMUserAttributes()
        
        if ((email) != nil) {
            attributes.email = email
            DispatchQueue.main.async {
                Intercom.loginUser(with: attributes) { result in
                    switch result {
                    case .success: call.resolve()
                    case .failure(let error): call.reject("Error logging in: \(error.localizedDescription)")
                    }
                }
            }
        }
        
        if ((userId) != nil) {
            attributes.userId = userId
            DispatchQueue.main.async {
                Intercom.loginUser(with: attributes) { result in
                    switch result {
                    case .success: call.resolve()
                    case .failure(let error): call.reject("Error logging in: \(error.localizedDescription)")
                    }
                }
            }
        }
    }
    
    @objc func loginIdentifiedUser(_ call: CAPPluginCall) {
        let userId = call.getString("userId")
        let email = call.getString("email")
        let attributes = ICMUserAttributes()
        
        if ((email) != nil) {
            attributes.email = email
            DispatchQueue.main.async {
                Intercom.loginUser(with: attributes) { result in
                    switch result {
                    case .success: call.resolve()
                    case .failure(let error): call.reject("Error logging in: \(error.localizedDescription)")
                    }
                }
            }
        }
        
        if ((userId) != nil) {
            attributes.userId = userId
            DispatchQueue.main.async {
                Intercom.loginUser(with: attributes) { result in
                    switch result {
                    case .success: call.resolve()
                    case .failure(let error): call.reject("Error logging in: \(error.localizedDescription)")
                    }
                }
            }
        }
    }
    
    @available(*, deprecated, message: "This method is deprecated, use loginIdentifiedUser() instead.")
    @objc func registerUnidentifiedUser(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Intercom.loginUnidentifiedUser { result in
                switch result {
                case .success: call.resolve()
                case .failure(let error): call.reject("Error loggin in unidentified user: \(error.localizedDescription)")
                }
            }
        }
    }
    
    @objc func loginUnidentifiedUser(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            Intercom.loginUnidentifiedUser { result in
                switch result {
                case .success: call.resolve()
                case .failure(let error): call.reject("Error loggin in unidentified user: \(error.localizedDescription)")
                }
            }
        }
    }
    
    @objc func updateUser(_ call: CAPPluginCall) {
        let userAttributes = ICMUserAttributes()
        let userId = call.getString("userId")
        if (userId != nil) {
            userAttributes.userId = userId
        }
        let email = call.getString("email")
        if (email != nil) {
            userAttributes.email = email
        }
        let name = call.getString("name")
        if (name != nil) {
            userAttributes.name = name
        }
        let phone = call.getString("phone")
        if (phone != nil) {
            userAttributes.phone = phone
        }
        let languageOverride = call.getString("languageOverride")
        if (languageOverride != nil) {
            userAttributes.languageOverride = languageOverride
        }
        let customAttributes = call.getObject("customAttributes")
        if (customAttributes != nil) {
            userAttributes.customAttributes = customAttributes
        }
        
        if let company = constructCompany(call.getObject("company")) {
            userAttributes.companies = [company]
        } else {
            if let companies = call.getArray("companies") as? [JSObject] {
                if (!companies.isEmpty) {
                    let companyArray = companies.compactMap { c in
                        constructCompany(c)
                    }
                    userAttributes.companies = companyArray
                }
            }
        }
        
        DispatchQueue.main.async {
            Intercom.updateUser(with: userAttributes) { result in
                switch result {
                case .success:
                    call.resolve()
                case .failure(let error):
                    call.reject("Error updating user: \(error.localizedDescription)")
                }
            }
        }
    }
    
    @objc func logout(_ call: CAPPluginCall) {
        Intercom.logout()
        call.resolve()
    }
    
    @objc func logEvent(_ call: CAPPluginCall) {
        guard let eventName = call.getString("name") else {
            call.reject("name is missing or empty")
            return
        }
        
        let metaData = call.getObject("data")
        
        if let metaData = call.getObject("data") {
            Intercom.logEvent(withName: eventName, metaData: metaData)
        } else {
            Intercom.logEvent(withName: eventName)
        }
        call.resolve()
    }
    
    @objc func present(_ call: CAPPluginCall) {
        let spaceMapping: [String: Space] = [
            "help": .helpCenter,
            "messages": .messages,
            "home": .home
        ]
        let spaceString = call.getString("space", "")
        let space = spaceMapping[spaceString] ?? .home
        Intercom.present(space)
        call.resolve()
    }
    
    @available(*, deprecated, message: "This method is deprecated, use present() instead.")
    @objc func displayMessenger(_ call: CAPPluginCall) {
        Intercom.present()
        call.resolve()
    }
    
    @objc func displayMessageComposer(_ call: CAPPluginCall) {
        guard let initialMessage = call.getString("message") else {
            call.reject("Enter an initial message")
            return
        }
        Intercom.presentMessageComposer(initialMessage);
        call.resolve()
    }
    
    @available(*, deprecated, message: "This method is deprecated, use present() instead.")
    @objc func displayHelpCenter(_ call: CAPPluginCall) {
        Intercom.present(Space.helpCenter)
        call.resolve()
    }
    
    @objc func hideMessenger(_ call: CAPPluginCall) {
        Intercom.hide()
        call.resolve()
    }
    
    @objc func displayLauncher(_ call: CAPPluginCall) {
        Intercom.setLauncherVisible(true)
        call.resolve()
    }
    
    @objc func hideLauncher(_ call: CAPPluginCall) {
        Intercom.setLauncherVisible(false)
        call.resolve()
    }
    
    @objc func displayInAppMessages(_ call: CAPPluginCall) {
        Intercom.setInAppMessagesVisible(true)
        call.resolve()
    }
    
    @objc func hideInAppMessages(_ call: CAPPluginCall) {
        Intercom.setInAppMessagesVisible(false)
        call.resolve()
    }
    
    @objc func setUserHash(_ call: CAPPluginCall) {
        guard let hmac = call.getString("hmsc") else {
            call.reject("hmac is missing or empty. Read intercom docs and generate it.")
            return
        }
        
        Intercom.setUserHash(hmac)
        call.resolve()
        print("hmac sent to intercom")
    }
    
    @objc func setBottomPadding(_ call: CAPPluginCall) {
        if let value = call.getString("value"),
           let number = NumberFormatter().number(from: value) {
            Intercom.setBottomPadding(CGFloat(truncating: number))
            call.resolve()
            print("set bottom padding")
        } else {
            call.reject("enter a value for padding bottom")
        }
    }
    
    @objc func presentContent(_ call: CAPPluginCall) {
        guard let contentId = call.getString("contentId") else {
            call.reject("contentId not defined")
            return
        }

        let contentMapping: [String: Intercom.Content] = [
            "carousel": Intercom.Content.carousel(id: contentId),
            "survey": Intercom.Content.survey(id: contentId),
            "article": Intercom.Content.article(id: contentId)
        ]
        let contentTypeString = call.getString("contentType", "")
        guard let contentType = contentMapping[contentTypeString] else {
            call.reject("contentType not found")
            return
        }

        Intercom.presentContent(contentType)
        call.resolve()
    }
    
    @available(*, deprecated, message: "This method is deprecated, use presentContent() instead.")
    @objc func displayCarousel(_ call: CAPPluginCall) {
        if let carouselId = call.getString("carouselId") {
            Intercom.presentContent(Intercom.Content.carousel(id: carouselId))
            call.resolve()
        } else{
            call.reject("carouselId not provided.")
        }
    }
    
    @available(*, deprecated, message: "This method is deprecated, use presentContent() instead.")
    @objc func displayArticle(_ call: CAPPluginCall) {
        if let articleId = call.getString("articleId") {
            Intercom.presentContent(Intercom.Content.article(id: articleId))
            call.resolve()
        } else {
            call.reject("articleId not provided.")
        }
    }
    
    @objc func getUnreadConversationCount(_ call: CAPPluginCall) {
        let unreadCount = Intercom.unreadConversationCount()
        call.resolve(["unreadCount": unreadCount])
    }
    
    private func constructCompany(_ companyData: JSObject?) -> ICMCompany? {
        guard let company = companyData else { return nil }
        
        let companyAttributes = ICMCompany()
        
        companyAttributes.companyId = company["companyId"] as? String ?? ""
            
        let name = company["name"] as? String ?? nil
        if (name != nil) {
            companyAttributes.name = name
        }
            
        let createdAt = company["createdAt"] as? TimeInterval ?? 0
        if (createdAt != 0) {
            companyAttributes.createdAt = Date(timeIntervalSince1970: TimeInterval(createdAt))
        }
        
        let monthlySpend = company["monthlySpend"] as? NSNumber ?? nil
        if (monthlySpend != nil) {
            companyAttributes.monthlySpend = monthlySpend
        }
        
        let plan = company["plan"] as? String ?? nil
        if (plan != nil) {
            companyAttributes.plan = plan
        }
        
        let customAttributes = company["customAttributes"] as? [String: Any] ?? nil
        if (customAttributes != nil) {
            companyAttributes.customAttributes = customAttributes
        }
        
        return companyAttributes
    }
}

